tw-tasks-core/src/main/java/com/transferwise/tasks/TasksService.java

```diff
@@ -189,6 +193,47 @@ public boolean resumeTask(ResumeTaskRequest request) {
         });
   }
 
+  @Override
+  @EntryPoint(usesExisting = true)
+  @Transactional(rollbackFor = Exception.class)
+  public RescheduleTaskResponse rescheduleTask(RescheduleTaskRequest request) {
+    return entryPointsHelper.continueOrCreate(EntryPointsGroups.TW_TASKS_ENGINE, EntryPointsNames.RESCHEDULE_TASK,
+        () -> {
+          UUID taskId = request.getTaskId();
+          mdcService.put(request.getTaskId(), request.getVersion());
+
+          FullTaskRecord task = taskDao.getTask(taskId, FullTaskRecord.class);
+
+          if (task == null) {
+            log.debug("Cannot reschedule task '" + taskId + "' as it was not found.");
+            return new RescheduleTaskResponse().setResult(Result.NOT_FOUND).setTaskId(taskId);
+          }
+
+          mdcService.put(task);
+
+          long version = task.getVersion();
+
+          if (version != request.getVersion()) {
+            coreMetricsTemplate.registerFailedNextEventTimeChange(task.getType(), task.getNextEventTime(), request.getRunAfterTime());
+            log.debug("Expected version " + request.getVersion() + " does not match " + version + ".");
+            return new RescheduleTaskResponse().setResult(Result.NOT_FOUND).setTaskId(taskId);
+          }
+
+          if (task.getStatus().equals(WAITING.name())) {
+            if (!taskDao.setNextEventTime(taskId, request.getRunAfterTime(), version, WAITING.name())) {
+              coreMetricsTemplate.registerFailedNextEventTimeChange(task.getType(), task.getNextEventTime(), request.getRunAfterTime());
+              return new RescheduleTaskResponse().setResult(RescheduleTaskResponse.Result.FAILED).setTaskId(taskId);
+            } else {
+              coreMetricsTemplate.registerTaskRescheduled(null, task.getType());
```

Comment:
how do I get bucket here? or do we even need it? 🤔

Reply:
I think null bucket is ok here.